### PR TITLE
Lock screen orientation to landscape on mobile phones

### PR DESCRIPTION
**Pull request by Susko3**
Made after some thought in https://github.com/ppy/osu-framework/pull/4897#issuecomment-1026746382.
Relevant issue: https://github.com/ppy/osu/issues/12764.

Why `SensorLandscape` was chosen over `UserLandscape`: https://github.com/ppy/osu-framework/pull/4897/files/bb55f2009a5ac2be489f91ab1537f8f2cdd54c0c#r759511353.

Landscape makes sense for osu! as it's primarily designed for landscape displays.

### CODE REVIEW

**Comment by Susko3** [comment]
We can't just change screen orientation when certain overlays are open, but maybe a toggle "Allow portrait during non-gameplay" (defaulting to disabled) could work.

I'm still a firm believer that the user should have more granular control (and one that applies to the whole game). Maybe not all options from the proposed config enum, but a select few. Showing different options for tablets and phones is also something to consider.

You do bring up a good point about chat. Using a virtual keyboard in landscape is bad UX on phones.

**Comment by bdach** [comment]
i'm just going to mention that this was attempted before and rejected. see https://github.com/ppy/osu/pull/10347 and subsequently https://github.com/ppy/osu/pull/10348 which introduced the current `fullUser` orientation.

i'm not sure i have a clear opinion on this at this point.

**Comment by peppy** [comment]
At this point, I am willing to concede to limiting to landscape only on phones only if there's a majority of developers that sees this as the correct path forward. The edge cases where this is going to improve the user experience are limited to things like

- Having a device flat on the desk and after orienting to landscape, the sensor misfiring and rotating back to portrait while the user is playing.
- Playing the game while lying sideways.

Every other case should already be "automatically" handled correctly.

So I'm still not so sure, as I also intend osu! to be a utility app which can potentially be used to check on one's profile, chat, messages (even for PC-main players) and not strictly be about gameplay.

@smoogipoo thoughts on this?

**Comment by smoogipoo** [comment]
I don't have an opinion on this because I don't have an Android phone.

**Comment by peppy** [comment]
It would also apply to iPhones if we made the change.

**Comment by smoogipoo** [comment]
So all this means is that the default is landscape, but the user can change to portrait at any point, is that correct? Maybe I'm just not understanding the point of this.

**Comment by peppy** [comment]
The proposal is to lock the app to landscape permanently (allowing either orientation of landscape aka 180 degree rotation, but no portrait).

**Comment by bdach** [comment]
Given that the UI does not generally feel good in portrait yet, at least not on the screen sizes of devices I own (everything turns out being rather small, and designs are generally... designed for landscape - see song select) I would not be strongly opposed to always forcing landscape myself and eliminating that problem in a sense, at least for a little while. I don't feel that confident about it to just go and merge this myself though.

**Comment by frenzibyte** [comment]
I have pushed a companion change for iPhones to keep consistency while this is being discussed (still keeping iPad to support all orientations using the special [`UISupportedInterfaceOrientations~ipad`](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/iPhoneOSKeys.html#//apple_ref/doc/uid/TP40009252-SW10) key).

**Comment by smoogipoo** [comment]
Locking to landscape sounds fine for now.

**Comment by peppy** [comment]
@Susko3 is it possible to allow tablets to still run in portrait? not that it necessarily needs to be done if we apply this change, but am curious.

**Comment by peppy** [comment]
In addition, as everyone seems to agree I'm fine with pushing out this change, and revisiting at a point things behave better in portrait orientation on phones.

**Comment by Susko3** [comment]
> @Susko3 is it possible to allow tablets to still run in portrait? not that it necessarily needs to be done if we apply this change, but am curious.

https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes

Should be possible, but it's not a clear-cut `IsTablet` API, it involves calculating screen size.

We're probably going for `ScreenOrientation.FullUser` on tablets, to match iPads.

**Comment by peppy** [comment]
Basing the decision on the aspect ratio seems quite amicable if there's no standard. 4:3 would be a good cut-off I think.

File: osu.Android/OsuGameActivity.cs [hunk 1]
```diff
@@ -39,10 +42,10 @@ namespace osu.Android
     [IntentFilter(new[] { Intent.ActionView }, Categories = new[] { Intent.CategoryBrowsable, Intent.CategoryDefault }, DataSchemes = new[] { "osu", "osump" })]
     public class OsuGameActivity : AndroidGameActivity
     {
-        public const ScreenOrientation DEFAULT_ORIENTATION = ScreenOrientation.SensorLandscape;
-
         private static readonly string[] osu_url_schemes = { "osu", "osump" };
 
+        public ScreenOrientation ScreenOrientation = ScreenOrientation.Unspecified;
```
Comment (frenzibyte) [inline osu.Android/OsuGameActivity.cs:47]: Would call this `DefaultScreenOrientations` for clarity.

**Comment by Susko3** [comment]
Have added a rudimentary tablet check, using 600 dp as recommended by the Android docs.

The activity no longer has a default screen orientation value, and will use the system default one while it's loading in.

I'll move this `isTablet` logic to framework at some point. But this will need to consider folding phones (which change display size), etc.

**Review by peppy — APPROVED** [review]

**Comment by smileyhead** [comment]
Just pitching in a comment, but I feel like that portrait would definitely be preferred on Mania (whenever touch support gets implemented), and possibly CtB, too.

**Comment by Susko3** [comment]
I mentioned something similar here https://github.com/ppy/osu/issues/12764#issuecomment-840030021.
Please keep all comments in that thread.

**Comment by frenzibyte** [comment]
Let's rather move discussion about "supporting portrait back" in https://github.com/ppy/osu/discussions/16783 instead, rather than #12764 (also tempted to convert #12764 into a discussion thread as well because there's not really a decided direction for it yet as far as I've seen).

**Pull request merged**